### DISCUSSION TITLE
Rate limiting requests to Amazon AWS services

### DISCUSSION THREAD

**Discussion by lonix1** [App-vNext/Polly#2340] [Q&A] [answered]
I've found many examples of using Polly with `HttpClient`. However I'm using AWS' .NET SDK which includes a custom client.

The service I'm using  (SES) limits requests to 50,000 per day and 14 per second. I'd like to run as many requests in parallel as possible, without exceeding the limit, and without queuing (if I hit the limit, there should be backpressure rather than queuing, which is computationally expensive).

Are there any simple samples of using Polly with AWS, Azure, etc.?

If not, and I must start from scratch, am I right in assuming that I should investigate the "Rate Limiting Policy"? Does it handle the above use case alone, or should it be paired with something else?

**Comment by martincostello** [comment, marked as answer]
Can you share a link to a specific `HttpClient` example you're looking at?

There are some `HttpClient`-specific ways to integrate Polly that "hide" it from the consumer by hooking into HttpClientFactory and adding Polly into the HTTP request pipeline transparently. The Microsoft extensions for Polly use this approach, for example.

Apart from the above, `HttpClient` isn't special and _any_ code can be wrapped with Polly to do retries, circuit-breaking, rate-limiting, timeouts etc.

Our [documentation for rate-limiting](https://www.pollydocs.org/strategies/rate-limiter.html) should be applicable for AWS, Azure or anything else.

If you want to go down the HttpClient-specific route, you would need to tie the creation of AWS clients into the HttpClient pipeline so that instances are retrieved from HttpClientFactory, rather than through its own-built-in mechanism.

For example you could create a custom implementation of [`Amazon.Runtime.HttpClientFactory`](https://github.com/aws/aws-sdk-net/blob/ca4005485bca2295d2985250ac25e774f01e2c4d/sdk/src/Core/Amazon.Runtime/Pipeline/HttpHandler/_netstandard/HttpRequestMessageFactory.cs) that uses `IHttpClientFactory` internally, and pass this to AWSSDK clients. This would then allow any Polly functionality associated with `HttpClient` to be automatically applied to AWSSDK clients provided they use your adapter/bridge/wrapper.

**Reply by lonix1** [reply]
> Can you share a link to a specific HttpClient example you're looking at?

I was referring to [AmazonSimpleEmailServiceV2Client](https://github.com/aws/aws-sdk-net/blob/main/sdk/src/Services/SimpleEmailV2/Generated/_netstandard/AmazonSimpleEmailServiceV2Client.cs#L55) which seems to be a custom implementation; from what I can tell it doesn't even wrap `HttpClient`.

> Our documentation for rate-limiting should be applicable for AWS, Azure or anything else.

That seems to do what I need. I just wanted to be sure I'm going in the right direction. Many thanks for your advice!

**Reply by martincostello** [reply]
All AWSSDK classes derived from `AmazonServiceClient` use `HttpClient` internally. They eventually come from [here](https://github.com/aws/aws-sdk-net/blob/b691e46e57a3e24477e6a5fa2e849da44db7002f/sdk/src/Core/Amazon.Runtime/Pipeline/HttpHandler/_mobile/HttpRequestMessageFactory.cs#L164-L202) by default. There's just lots of hoops to jump through to stitch it together with HttpClientFactory.
